[Test Generation] Control inlining using method annotations

Dargones opened the pull request:
Fixes #3716

This is a second attempt at adding more fine grained control over inlining to test generation. In lieu of a command line option (as in #3728), this PR introduces a `:testInline` attribute which can be put on methods and functions. The attribute must be followed by a positive integer, which in the presence of recursion gives the bound on how many times the given method can be unrolled (1 means inlining but no unrolling). In addition to using the attribute, one must specify the `--inline` boolean command line option to activate inlining (because inlining requires changes to the parsing process which must happen before I can look at method attributes). All of this very closely mirrors Boogie's API for inlining, so I think it should be more intuitive than the previous solution.

@codyroux, @hmijail, @aleks-AWS, I would be grateful for any feedback on how well this addresses the needs of various applications since there are various trade-off involved in-between this and #3728! Below I discuss some of the details and give an example.

I added various error messages that the user gets if they do not specify the unrolling limit or if they use the `:testInline` attribute but not the `--inline` command line option. Using `--inline` with `--verbose` will also output the list of methods/functions that were inlined, so the user can double-check that the desired methods are in-fact being inlined.

One of the alternatives would be to make inlining behavior specific to the target tested method, as @keyboardDrummer proposes in the linked issue. My guess is that the use case for this is relatively uncommon as there is usually one specific high-level method that has to be targeted but it is true that the solution proposed in this PR is more restrictive. Is there an application that would require different inlining behavior based on what method is being tested?

As an example of how inlining would behave if this PR gets merged, consider the following Dafny program that recursively computes the remainder of dividing some number by `3` (this test case is added as part of the PR):

```Dafny
module M {
  class Inlining {
    function {:testInline 2} mod3 (n:int):int 
      requires n >= 0
      decreases n
    {
      if n == 0 then 0 else
      if n == 1 then 1 else
      if n == 2 then 2 else
      mod3(n-3)
    }
    method test(n:int) returns (r:int) 
      requires n >= 3
    {
      r := mod3(n);
    }
  }
}
```

To generate tests for this program, one could execute:
```bash
dafny generate-tests Block --inline --target-method:M.Inlining.test FILENAME.dfy --verbose
```

Because the value of `:testInline` is `2`, test generation would produce at least `3` tests, one for each of the base cases in the recursive `mod3` function. With `:testInline` set to `1`, test generation would not be able to see beyond the first call to `mod3`, and so it would not be able to generate tests for the base cases as they are effectively unreachable without a recursive call.

<!-- Is this a user-visible change?  Remember to update RELEASE_NOTES.md -->

<!-- Is this a bug fix?  Remember to include a test in Test/git-issues/ -->

<!-- Is this a bug fix for an issue introduced in the latest release?  Mention this in the PR details and ensure a patch release is considered -->

<!-- Does this PR need tests?  Add them to `Test/` or to `Source/*.Test/…` and run them with `dotnet test` -->

<!-- Are you moving a large amount of code? Read CONTRIBUTING.md to learn how to do that while maintaining git history -->

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

keyboardDrummer commented:
> One of the alternatives would be to make inlining behavior specific to the target tested method, as @keyboardDrummer proposes in the linked issue. My guess is that the use case for this is relatively uncommon as there is usually one specific high-level method that has to be targeted but it is true that the solution proposed in this PR is more restrictive. Is there an application that would require different inlining behavior based on what method is being tested?

That could be a future feature instead of an alternative, `{:testInline M.Inlining 3}` on the root method could override the `{:testInline 2}` on the `M.Inlining` method.

keyboardDrummer commented:
> (because inlining requires changes to the parsing process which must happen before I can look at method attributes)

I'm not sure which parsing process you're referring to. Looking at where `TestGenOptions.TestInline` is used, that seems to be way after parsing the Dafny program.

keyboardDrummer commented on Source/DafnyCore/TestGenerationOptions.cs:
```diff
@@ -52,11 +53,8 @@ public enum Modes { None, Block, Path };
-          if (ps.GetIntArgument(ref depth)) {
-            TestInlineDepth = (uint)depth;
-          }
+        case "generateTestInline":
```
Consider using `DafnyOptions.RegisterLegacyUi` to replace the `generateTestInline` related code in this file.

keyboardDrummer reviewed (commented):
Looks great, left some minor comments

Dargones commented:
Thank you, @keyboardDrummer for all the feedback! I have removed the `--inline` option like you suggest above (meaning that I parse the input program, scan it for the presence of the `:testInline` attribute and then re-parse it again if the attribute is present). The re-parsing is necessary because I need to look at the bodies of all inlined methods/functions, even if they are defined in a different file (so I have to parse with `/verifyAllModules` flag).

I think this should also address your comments about how to register the command line option properly (since I removed the option altogether), but I am wondering if I should make a separate PR to make the other options use the `DafnyOptions.RegisterLegacyUi` functionality. Would you recommend creating a `TestGenerationOptionsBag` file for this purpose?

Dargones commented on Source/DafnyTestGeneration/GenerateTestsCommand.cs:
```diff
@@ -84,8 +84,11 @@ private enum Mode {
     });
-    DafnyOptions.RegisterLegacyBinding(TestInlineDepth, (options, value) => {
-      options.TestGenOptions.TestInlineDepth = value;
+    DafnyOptions.RegisterLegacyBinding(TestInline, (options, value) => {
```
I have decided to remove this option completely and just scan the program for the `:testInline` attribute and reparse if necessary (see my comment below), but thank you for clarifying how to register the options properly, this is very helpful!

Dargones commented:
Have made a commit that ensures the transformations I apply to support inlining only modify procedures that are going to be inlined - this should significantly improve test generation performance.

atomb commented:
It looks as though some code changed on `master` removed or moved a field this PR is using.

Dargones commented:
Thank you, @atomb! Should work now
